caffe2/operators/mean_op.cc

```diff
@@ -79,19 +79,21 @@ print("Mean:", workspace.FetchBlob("X"))
 class GetMeanGradient : public GradientMakerBase {
   using GradientMakerBase::GradientMakerBase;
   vector<OperatorDef> GetGradientDefs() override {
+    auto inputs = std::vector<string>();
     auto outputs = std::vector<string>();
+    inputs.push_back(GO(0));
     for (int i = 0; i < def_.input_size(); i++) {
+      inputs.push_back(I(i));
       outputs.push_back(GI(i));
     }
-    return SingleGradientDef(
-        "MeanGradient", "", std::vector<string>{GO(0)}, outputs);
+    return SingleGradientDef("MeanGradient", "", inputs, outputs);
   }
 };
 
 REGISTER_GRADIENT(Mean, GetMeanGradient);
 
 OPERATOR_SCHEMA(MeanGradient)
-    .NumInputs(1)
+    .NumInputs(1, INT_MAX)
```

Comment:
In the gradient op, you are propagating the gradients back. The input should be GO(0), and you only have one output for Mean op. So for MeanGradient, the input number should 1 as well.

Reply:
The reason I'm adding extra inputs is to obtain the original sizes of the input tensors. I couldn't find another way to do this other than passing `I(0), I(1)`, etc as part of the input to the gradient operator (as far as I can tell, `GI(0)...GI(N)` seem to be all set by default to the fully broadcasted size). Note that this is similar to the [gradient of the elementwise_add operator](https://github.com/pytorch/pytorch/blob/d47108add0eaddbdaf1cf8bc95599faac66442cd/caffe2/operators/elementwise_add_gradient_op.cc#L24), which receives the two inputs even though the add operation only has one output. 

I'm happy to change this if there is a better way to do it, just not sure how that would be done.

Reply:
That's a good point, it seems currently we don't have a better solution for the input sizes.